### PR TITLE
Turn chroma-common into a python package

### PR DESCRIPTION
**Pull request by tanabarr**
Currently chroma-common is not a python package and other packages
link to it through symbolic links (chroma-agent and chroma-manager).
This is inconvenient for a number of reasons including having to have
unit tests for chroma-common reside in directories other than
chroma-common and IDEs not be to resolve the links to common code
which adds inconvenience during development.

Change chroma-common into a stand-alone python package and distribute
alongside the other IML packages. This provides separation of the
common code and an ability to test in isolation and enables code to
import the package directly rather than through symbolic links.

  - Create python package dir structure for chroma-common with
    boilerplate files
  - Remove symlinks to chroma_common
  - Update imports to reference chroma_common package directly
    rather than symlinks to chroma-common dir
  - Build as a package containing source but no install scripts
  - Add chroma_common directory to sys.path in
    update_chroma-externals.py
  - Add __init__.py to chroma-common/chroma_common
  - Remove __init__.py from chroma-common/
  - Make and install common package on builder before manager
    to satisfy import dependencies when building docs
  - Add chroma_common directory to sys.path in chroma-agent/Makefile
  - Add chroma_common directory to sys.path in tests framework for
    unit-tests and services where needed including log collection
  - Ensure package file name of chroma-common are consistent with
    chroma-manager, e.g. ...elX.<ARCH>.rpm
  - Tidy chroma package dependency tree in spec files
  - Add "make -C ../chroma-common develop" last to manager target
    "make develop"
  - Set chroma_common/lib/util.py _use_threads_default to True as
    we only want Manager unit tests not to use threads because
    unittest runs within a database transaction
  - Explicitly set _use_threads_default to False in IMLUnitTestCase
    which is then applied to all manager unit tests
  - Use print function syntax in behave tests to be compatible
    with up-to-date versions of python behave

### CODE REVIEW

**Comment by brianjmurrell** [comment]
Is this a duplicate of #26?

**Comment by tanabarr** [comment]
Brian: yes I will close #26 your comments have been addressed in this patch

**Comment by jgrund** [comment]
How are the tests looking with this change?

**Comment by jgrund** [comment]
Can we remove the Change-Id for patches that come over from gerrit?

File: Makefile [hunk 1]
```diff
@@ -42,8 +42,8 @@ deps: repo
 
 tags:
 	#find chroma-agent/chroma_agent chroma-manager/{tests,chroma_{agent_comms,api,cli,core,ui}} -type f | ctags -L -
-	ctags --python-kinds=-i -R --exclude=chroma-manager/_topdir --exclude=chroma-\*/myenv\* --exclude=chroma-dependencies --exclude=chroma_unit_test_env --exclude=chroma-externals --exclude=chroma-manager/ui-modules .
+	ctags --python-kinds=-i -R --exclude=chroma-manager/_topdir --exclude=chroma-\*/myenv\* --exclude=chroma-common --exclude=chroma-dependencies --exclude=chroma_unit_test_env --exclude=chroma-externals --exclude=chroma-manager/chroma_ui* exclude=chroma-manager/ui-modules .
```
Comment (jgrund) [inline Makefile]: chroma-ui is no more.

File: chroma-agent/tests/common/blockdevices/test_blockdevice_linux.py [hunk 2]
```diff
@@ -1,7 +1,7 @@
 import mock
 
-from chroma_agent.chroma_common.blockdevices.blockdevice_linux import BlockDeviceLinux
-from tests.chroma_common.blockdevices.blockdevice_base_tests import BaseTestBD
+from chroma_common.blockdevices.blockdevice_linux import BlockDeviceLinux
+from tests.common.blockdevices.blockdevice_base_tests import BaseTestBD
```
Comment (jgrund) [inline chroma-agent/tests/common/blockdevices/test_blockdevice_linux.py]: Why is this renamed from chroma_common?

File: chroma-common/LICENSE.txt [hunk 3]
```diff
@@ -0,0 +1,7 @@
+Copyright (c) 2016 Intel Corproration. All Rights Reserved.
```
Comment (jgrund) [inline chroma-common/LICENSE.txt]: License needs to be fixed.

We should remove this file until it lives in it's own package. At which point it would make sense to add the MIT license.

File: chroma-common/README.txt [hunk 4]
```diff
@@ -0,0 +1,29 @@
+#
```
Comment (jgrund) [inline chroma-common/README.txt]: Remove license, also not current license.

File: chroma-common/Makefile [hunk 5]
```diff
@@ -0,0 +1,40 @@
+include ../include/Makefile.version
```
Comment (jgrund) [inline chroma-common/Makefile]: Since this is being extracted out, we should also move it out into it's own package.

File: chroma-common/chroma_common/__init__.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+# licensors. The Material is protected by worldwide copyright and trade secret
+# laws and treaty provisions. No part of the Material may be used, copied,
+# reproduced, modified, published, uploaded, posted, transmitted, distributed,
+# or disclosed in any way without Intel's prior express written permission.
+#
+# No license under any patent, copyright, trade secret or other intellectual
+# property right is granted to or conferred upon you by disclosure or delivery
+# of the Materials, either expressly, by implication, inducement, estoppel or
+# otherwise. Any license under such intellectual property rights must be
+# express and approved by Intel in writing.
```
Comment (jgrund) [inline chroma-common/chroma_common/__init__.py]: License incorrect.

File: chroma-common/chroma_common/lib/util.py [hunk 7]
```diff
@@ -59,10 +59,10 @@ def insert(self, index, value):
...
 
 # When running unit tests every test is within a transaction and so if you kick of another thread you will not see
-# any of the database changes that have occurred. For this reason if we are running unit tests we default to no threads
-# otherwise we default to threads.
-# Always run threads on the agent because the agent doesn't use the db.
-_use_threads_default = (not running_nose_tests()) or ('chroma_agent' in __name__)
+# any of the database changes that have occurred. Threads will be enabled by default but manager unit tests must
+# explicitly set _use_threads_default to False to avoid database related failures. Agent unit tests should always
+# run threads on the agent because the agent doesn't use the db.
+_use_threads_default = True
```
Comment (jgrund) [inline chroma-common/chroma_common/lib/util.py]: This looks unrelated to the patch. Is this from a merge conflict?

File: chroma-manager/tests/feature/cli/features/steps/standard_cli.py [hunk 8]
```diff
@@ -3,6 +3,7 @@
 # Copyright (c) 2012 Whamcloud, Inc.  All rights reserved.
 # ========================================================
 
+from __future__ import print_function
```
Comment (jgrund) [inline chroma-manager/tests/feature/cli/features/steps/standard_cli.py]: Why is this being added? Also, header needs updating.

File: chroma-manager/tests/unit/lib/iml_unit_test_case.py [hunk 9]
```diff
@@ -2,7 +2,10 @@
 
 from chroma_core.models import Command
 from chroma_core.services.log import log_register
+from chroma_common.lib import util
 
+# Manager unit tests should not use threading, see comment in util.py
+util._use_threads_default = False
```
Comment (jgrund) [inline chroma-manager/tests/unit/lib/iml_unit_test_case.py]: Is this related to this patch?

Comment (tanabarr) [inline chroma-manager/tests/unit/lib/iml_unit_test_case.py]: yes because we have split out AgentUnitTestCase from ImlUnitTestCase and we don't want tests inheriting from the latter using threads by default, whereas we do for the former

File: chroma-manager/tests/feature/cli/features/steps/standard_cli.py (hunk 8 above)
Comment (tanabarr) [inline chroma-manager/tests/feature/cli/features/steps/standard_cli.py]: in order to print the debug necessary to diagnose and fix an issue related to the changes in this patch in the behave tests, this had to be added (behave wasn't working with the original print function when the AssertionError was raised). I will update the header

File: chroma-common/Makefile (hunk 5 above)
Comment (brianjmurrell) [inline chroma-common/Makefile]: I also have a branch where I have started this kind of (separation of the Chroma python modules) work.  We should make sure not to duplicate each others' work.

Comment (tanabarr) [inline chroma-common/Makefile]: @brianjmurrell this patch has been around on gerrit for a number of months and there are 2 follow-on patches that have similarly been around for a while, probably best to check you are not duplicating them with similar work.

Comment (brianjmurrell) [inline chroma-common/Makefile]: @tanabarr: Sorry, I was not clear.  The patch I have it not about extracting chroma-common (or any other ingrained bits) out of overlapping modules but rather about extracting the various already defined packages such as chroma-manager, chroma-agent, chroma-diagnostics, etc. out of a single "project" into distinct projects.

The general guideline we should be moving toward is one RPM specfile per project rather than the current 3-4 that we have in the "chroma" project.

File: chroma-manager/tests/feature/cli/features/steps/standard_cli.py (hunk 8 above)
Comment (tanabarr) [inline chroma-manager/tests/feature/cli/features/steps/standard_cli.py]: have removed the header (test code where similar files have no header) and removed the changes

File: chroma-manager/Makefile [hunk 10]
```diff
@@ -221,8 +224,8 @@ requirements.txt: ../chroma-dependencies/Makefile ../chroma-dependencies/*/Makef
 	echo "# DO NOT UPDATE THIS FILE MANUALLY.  See the warning at the top of this file."; \
 	sort < requirements.tmp; } > $@
 	rm -f requirements.tmp
-	set -e;                                                   \
-	if grep ^make $@; then                                    \
+	set -e;                                               \
+	if grep ^make $@; then                                \
```
Comment (brianjmurrell) [inline chroma-manager/Makefile]: The above two lines' line continuators should line up with the ones below.

File: chroma-manager/Makefile [hunk 11]
```diff
@@ -235,14 +238,15 @@ download: requirements.txt
 
 docs: requirements.txt version
 	echo "jenkins_fold:start:Make Docs"
-	set -e;                                                         \
-	if [ -z $(VIRTUAL_ENV) ] ; then                                 \
+	set -e;                                                     \
+	export PYTHONPATH=$$(pwd)/../chroma-common:$$PYTHONPATH;    \
+	if [ -z $(VIRTUAL_ENV) ] ; then                             \
```
Comment (brianjmurrell) [inline chroma-manager/Makefile]: Ditto with the line continuators in this block including L249 below.

File: chroma-manager/tests/feature/cli/features/steps/standard_cli.py [hunk 12]
```diff
@@ -1,8 +1,3 @@
-#
-# ========================================================
-# Copyright (c) 2012 Whamcloud, Inc.  All rights reserved.
-# ========================================================
-
```
Comment (brianjmurrell) [inline chroma-manager/tests/feature/cli/features/steps/standard_cli.py]: Does this need to be replaced with our new copyright header rather than just deleted?

File: chroma-common/Makefile (hunk 5 above)
Comment (tanabarr) [inline chroma-common/Makefile]: seeing as this work is already done and will verify that we can separate it out whilst keeping necessary functionality in place and without introducing bugs I would prefer to land this first

File: chroma-common/chroma-common.spec [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+%{?!version: %define version %(%{__python2} -c "from chroma_common import version; sys.stdout.write(version())")}
+%{?!package_release: %define package_release 1}
+%{?!python2_sitelib: %define python2_sitelib %(%{__python2} -c "from distutils.sysconfig import get_python_lib; import sys; sys.stdout.write(get_python_lib())")}
+
+Summary: Chroma Common
+Name: %{name}
+Version: %{version}
+Release: %{package_release}%{?dist}
+Source0: %{name}-%{version}.tar.gz
+License: Proprietary
```
Comment (tanabarr) [inline chroma-common/chroma-common.spec]: Licence is a requirement here and given we don't / haven't decided on what to change all the spec files to, will keep it consistent with the others and change them all in one patch

File: chroma-manager/tests/feature/cli/features/steps/standard_cli.py (hunk 12 above)
Comment (tanabarr) [inline chroma-manager/tests/feature/cli/features/steps/standard_cli.py]: it is test code and files in the same directory (as with most test files) do not have Copyright headers. if we want to change this we should do it in a separate patch

File: chroma-agent/Makefile [hunk 14]
```diff
@@ -41,12 +43,13 @@ rpms: production cleandist tarball
...
-		 --define "version $(PACKAGE_VERSION)"         \
-		 --define "package_release $(PACKAGE_RELEASE)" \
-		 --define "%dist $$dist"                       \
+	export PYTHONPATH=$$PYTHONPATH:$$(pwd)/../chroma-common;  \
+	dist=$$(rpm --eval %dist);                                \
+	dist=$${dist/.centos/};                                   \
+	rpmbuild --define "_topdir $$(pwd)/_topdir"               \
+		 --define "version $(PACKAGE_VERSION)"                \
+		 --define "package_release $(PACKAGE_RELEASE)"        \
+		 --define "%dist $$dist"                              \
```
Comment (tanabarr) [inline chroma-agent/Makefile]: corrected in https://github.com/intel-hpdd/intel-manager-for-lustre/pull/76

File: chroma-manager/Makefile (hunk 10 above)
Comment (tanabarr) [inline chroma-manager/Makefile]: corrected in https://github.com/intel-hpdd/intel-manager-for-lustre/pull/76

File: chroma-manager/Makefile (hunk 11 above)
Comment (tanabarr) [inline chroma-manager/Makefile]: corrected in https://github.com/intel-hpdd/intel-manager-for-lustre/pull/76
